src/uu/dmesg/src/dmesg.rs

```diff
@@ -0,0 +1,151 @@
+
+mod json;
+
+const ABOUT: &str = help_about!("dmesg.md");
+const USAGE: &str = help_usage!("dmesg.md");
+
+#[uucore::main]
+pub fn uumain(args: impl uucore::Args) -> UResult<()> {
+    let mut dmesg = Dmesg::new();
+    let matches: clap::ArgMatches = uu_app().try_get_matches_from(args)?;
+    if let Some(kmsg_file) = matches.get_one::<String>(options::KMSG_FILE) {
+        dmesg.kmsg_file = kmsg_file;
+    }
+    if matches.get_flag(options::JSON) {
+        dmesg.output_format = OutputFormat::Json;
+    }
+    dmesg.parse()?.print();
+    Ok(())
+}
+
+pub fn uu_app() -> Command {
+    Command::new(uucore::util_name())
+        .override_usage(format_usage(USAGE))
+        .about(ABOUT)
+        .version(crate_version!())
+        .arg(
+            Arg::new(options::KMSG_FILE)
+                .short('K')
+                .long("kmsg-file")
+                .help("use the file in kmsg format")
+                .action(ArgAction::Set),
+        )
+        .arg(
+            Arg::new(options::JSON)
+                .short('J')
+                .long("json")
+                .help("use JSON output format")
+                .action(ArgAction::SetTrue),
+        )
+}
+
+mod options {
+    pub const KMSG_FILE: &str = "kmsg-file";
+    pub const JSON: &str = "json";
+}
+
+struct Dmesg<'a> {
+    kmsg_file: &'a str,
+    output_format: OutputFormat,
+    records: Option<Vec<Record>>,
+}
+
+impl Dmesg<'_> {
+    fn new() -> Self {
+        Dmesg {
+            kmsg_file: "/dev/kmsg",
+            output_format: OutputFormat::Normal,
+            records: None,
+        }
+    }
+
+    fn parse(mut self) -> UResult<Self> {
+        let mut records = vec![];
+        let re = Self::record_regex();
+        let lines = self.read_lines_from_kmsg_file()?;
+        for line in lines {
+            for (_, [pri_fac, seq, time, msg]) in re.captures_iter(&line).map(|c| c.extract()) {
+                records.push(Record::from_str_fields(pri_fac, seq, time, msg.to_string()));
+            }
+        }
+        self.records = Some(records);
+        Ok(self)
+    }
+
+    fn record_regex() -> Regex {
+        let valid_number_pattern = "0|[1-9][0-9]*";
+        let additional_fields_pattern = ",^[,;]*";
+        let record_pattern = format!(
+            "(?m)^({0}),({0}),({0}),.(?:{1})*;(.*)$",
+            valid_number_pattern, additional_fields_pattern
+        );
+        Regex::new(&record_pattern).expect("invalid regex.")
+    }
+
+    fn read_lines_from_kmsg_file(&self) -> UResult<Vec<String>> {
+        let mut lines = vec![];
+        let mut line = vec![];
+        let kmsg_bytes = fs::read(self.kmsg_file)
+            .map_err_context(|| format!("cannot open {}", self.kmsg_file))?;
+        for byte in kmsg_bytes {
+            if byte == 0 {
+                lines.push(String::from_utf8_lossy(&line).to_string());
+                line.clear();
+            } else {
+                line.push(byte);
+            }
+        }
```

Comment:
I think you can improve this code and get rid of the manual handling of `line` by using a more functional approach:
```suggestion
        let kmsg_bytes = fs::read(self.kmsg_file)
            .map_err_context(|| format!("cannot open {}", self.kmsg_file))?;
        let lines = kmsg_bytes
            .split(|&byte| byte == 0)
            .map(|line| String::from_utf8_lossy(line).to_string())
            .collect();
```

Reply:
Ah that's way cleaner, thanks! I applied the changes in commit df3fb06.